Support initialisation of footer.md

Chng-Zhi-Xuan opened the pull request:
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [X] Documentation update
• [X] New feature
• [X] Bug fix

<!--
    If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"
-->
Resolves #236
Fixes #241 
<!--
    Please ensure your pull request is ready:
    - Bug fix PR that is non-trivial **should** add a page or unit test for regression testing.
    - Feature PR **must** add a page to the user guide for demo.
    - Enhancement PR **should** update the user guide.

    Otherwise, prefix your PR title with "[WIP]".
-->

**What is the rationale for this request?**
User would like to have a default footer.md generated when initialising MarkBind.

**What changes did you make? (Give an overview)**

Functionality | Code 
--- | --- 
Generated footer.md in `_markbind/footers` directory <br>`Index.md` now specifies `footer.md` in `Front Matter` by default during `init`|  file generated in `initSite` and add default values for `footer.md`  
Support for a dynamic height footer using CSS Flexbox | To enable this, an empty `div` with the appropriate `#flex-div` CSS needs to be appended before a footer. This is done through `insertFlexDivBeforeLastFooter` , called after `insertTemplates` 
Remove duplicate footers in a single page. Due to being able to include other pages (which may contain a footer of their own) | Duplicates removed in new util function `removeDuplicateFooter`, called after `insertFlexDivBeforeLastFooter`
Ensure footers are not enclosed in any other component or html tag | Done in util function `unwrapFooter`, called after `removeDuplicateFooter`
Bug fix for #241 | `decodeURIComponent` for filePath in `_preprocess`


Demo of feature: |
--- |
**Note:** Background colours are modified to emphasise different parts of the components. It will be reverted for final review. <br><br>Green: app container which houses all generated content (appContents + footer) <br>Beige: Portion of `<div id="appContents">`, to enable dynamic height footer <br> Blue: Portion of footer, with intentional negative margins so it extends to cover the background of the app container. <br> ![footer-v2-demo](https://user-images.githubusercontent.com/31084833/40415533-ab99f236-5ead-11e8-9a6c-dea505379812.gif) <br> | 

Footer styling |
--- |
![footerstyleone](https://user-images.githubusercontent.com/31084833/40415500-932edf0e-5ead-11e8-9ad4-f353ec7b2124.PNG)

**Is there anything you'd like reviewers to focus on?**
- Default content for footer.md / index.md
- Variable / Method naming
- Updated Unit Tests
- User Guide wording 

**Testing instructions:**
1. Run `markbind init` on a blank folder.
2.  `markbind serve` should generate the footer correctly (bottom of page) despite sparse content.
3. Add long content to same page
4. Footer should be at the bottom of the page after scrolling
5. Define own custom `<footer>` in source lines. Delete `footer: footer.md` in `Front Matter` The custom footer should be the only one that renders.
6. Create another page that specifies a footer in its `Front Matter` or has a custom `<footer>`.
7. Include new page within original page. 
8. Included page should have contents generated but not its footer, on the original page.

Chng-Zhi-Xuan commented:
Would like comments on what should be the CSS and default values of footer.md. 

Tentatively I am thinking of a `light-grey` background, and black text. Along with modifying the `href text` colour to be the blue on our logo.

damithc commented:
> Would like comments on what should be the CSS and default values of footer.md.

Something that catch less attention than normal text would be good. May be you can screenshot some options?

yamgent commented:
```html
<include src=C:\Users\Tan Wang Leng\Desktop\markbind-projects\pr-238\footer.md />
```

As discussed, I did step 1 & 2, and the generated footer doesn't render properly (because the file path used is the entire file path, and it happens to have spaces). Ideally, we would not want the entire path to be appended in (if there are multiple authors, other users may have a different path as well).

Also need double quote for the src attribute.

yamgent commented:
In the generated `index.md`:

```html
<div id="appContents">
<frontmatter>
title: "Hello World"
</frontmatter>

# Hello world
Welcome to your page generated with MarkBind.
</div>

<include src="footer.md" />
```

The addition of `<div id="appContents">` will make the markdown file very cluttering to the user (they might not understand why the div is in there, and it is probably not necessary for the user to know that the content is wrapped around a div). Ideally these divs should be generated by MarkBind during processing, rather than explicitly stated inside the markdown files.

If there's no easy way for MarkBind wrap around the content with the div during processing, perhaps we should rethink the way we include our `footer.md`?

Chng-Zhi-Xuan commented:
**Update**
- PR contents updated.
- Changed behaviour of code such that it does not perform any write operations to either index.md or footer.md files. 
- Changed User defined footers taking precedence to last footer in file taking precedence.

yamgent commented on asset/css/markbind.css:
```diff
@@ -40,3 +40,26 @@ blockquote {
+    margin-top: 40px;
+    margin-right: -15px;
+    margin-left: -15px;
+}
```
Remember to add a newline at the end of this file

yamgent commented on lib/Page.js:
```diff
@@ -62,10 +63,43 @@ function calculateNewBaseUrl(filePath, root, lookUp) {
   return calculate(filePath, []);
 }
 
+function insertFlexDiv(pageData) {
```
`insertFlexDivBeforeFooter` might be more descriptive.

Chng-Zhi-Xuan replied:
Changed behaviour of function to be more like `insertFlexDivBeforeLastFooter`

yamgent commented on lib/Page.js:
```diff
@@ -62,10 +63,43 @@ function calculateNewBaseUrl(filePath, root, lookUp) {
   return array.filter((item, pos, self) => self.indexOf(item) === pos);
 }
 
+function removeDuplicateFooter(pageData) {
```
`removeDuplicateFooters` (end with 's' because the method is capable of removing more than one footer).

yamgent commented on asset/css/markbind.css:
```diff
@@ -40,3 +40,26 @@ blockquote {
+    min-height: 100vh;
+}
+
+#flexDiv{
```
Add a space before `{`, i.e.

```
#flexDiv {
```

yamgent commented on asset/css/markbind.css:
```diff
@@ -40,3 +40,26 @@ blockquote {
+
+footer {
+    color: black;
+    background-color: #ebebeb;
```
If we are keeping the colors, remember to sort the attributes according to alphabetical order.

yamgent commented on lib/Page.js:
```diff
@@ -62,10 +63,43 @@ function calculateNewBaseUrl(filePath, root, lookUp) {
+  const $ = cheerio.load(pageData);
+  const footers = $('footer');
+
+  footers.each((index, footer) => {
```
If we are only using one footer in the end, seems a bit redundant to add flexDiv to all footers. :/

yamgent commented on lib/Site.js:
```diff
@@ -158,7 +171,7 @@ Site.initSite = function (rootPath) {
           return Promise.resolve();
         }
-        return fs.outputFileAsync(indexPath, INDEX_MARKDOWN_DEFAULT);
+        return fs.outputFileAsync(indexPath, `${INDEX_MARKDOWN_DEFAULT}\n<include src="${FOOTER_FILE}" />`);
```
I think can just modify `INDEX_MARKDOWN_DEFAULT` to include the new footer.

Chng-Zhi-Xuan replied:
Leftover from previous implementation that used `rootPath`, will refactor this.

yamgent commented on lib/Page.js:
```diff
@@ -62,10 +63,43 @@ function calculateNewBaseUrl(filePath, root, lookUp) {
+    footers.remove();
+    flexDivs.last().after(lastFooter);
+  }
+  if (flexDivs.last().parents().length > 0) {
```
~~Add a comment to explain what this portion of the code is doing (something along the line of removing the divs that were wrapping the footer).~~ Or it can be a separate method on its own. EDIT: Think it should be a separate method, since this method is only concerned with removing duplicate footers.

yamgent replied:
As you mentioned, also handle the case where the user does this:

```
<div>
  <div>
    <div>
      <div>
        <footer>
          The user wrapped so many div inside the footer, but shouldn't.
        </footer>
      </div>
    </div>
  </div>
</div>
```

Chng-Zhi-Xuan replied:
Handling a case like that would require a `while loop` that iteratively destroys parent html tags until it has no more parents.

Chng-Zhi-Xuan commented:
**Update** 

- Found a formatting issue when using `Cheerio` after `markbinder.renderFile`. `Cheerio` formats redundant or single html tags to end with `/>` .

HTML loaded | HTML output from Cheerio 
--- | --- 
`<panel></panel>` | `<panel />`
`<img src="...">` | `<img src="..." />`

It also affects formatting of the `glyphicons` and text nearby 

`glyphicons` before | `glyphicons` after
--- | ---
`<span class=... ></span>` | `<span class=.... />`

- Work around is to include blank content within `glyphicon` span `&#8203`. 
(Take note of the spacing in the text)

No blank content | 
--- | 
![footerstyleone](https://user-images.githubusercontent.com/31084833/40469284-98f10b32-5f62-11e8-8f8a-79204e52311d.PNG) |  

includes `&#8203` |
--- |
![correct-footer-format](https://user-images.githubusercontent.com/31084833/40469296-9ef8716e-5f62-11e8-9deb-6056113bc1c2.PNG)

yamgent commented on lib/Site.js:
```diff
@@ -60,11 +61,23 @@ const SITE_CONFIG_DEFAULT = {
   + '# Hello world\n'
-  + 'Welcome to your page generated with MarkBind.\n';
+  + 'Welcome to your page generated with MarkBind.\n'
+  + `\n<include src="${FOOTER_FILE}" />`;
```
Put the newline character in the previous line.

yamgent commented on docs/userGuide/contentAuthoring.md:
```diff
@@ -55,6 +55,40 @@ The HTML output will be:
+
+You can use `footer.md` to standardize footers between pages. 
+
+In each page, you have to [include](includingContents.html) it manually for it to render. 
```
you have to include **footer.md** manually for it to render.

yamgent commented on asset/css/markbind.css:
```diff
@@ -40,3 +40,26 @@ blockquote {
+    min-height: 100vh;
+}
+
+#flexDiv {
```
ID naming convention: should be `flex-div` rather than `flexDiv`.

yamgent commented on lib/Page.js:
```diff
@@ -62,10 +63,55 @@ function calculateNewBaseUrl(filePath, root, lookUp) {
+function removeFooterParentTags(pageData) {
+  const $ = cheerio.load(pageData);
+  const lastFooter = $('footer').last();
+  const lastFooterParentNum = lastFooter.parents().length;
```
Does `lastFooter.parent()` give you the nested HTML elements? If so, is it possible to just go to the highest layer and replace that?

Chng-Zhi-Xuan replied:
Iterative method changed to constant time method:
```javascript
  if (lastFooterParentNum > 0) {
    const lastParentOfFooter = lastFooter.parents().last();
    const footerParentContents = lastFooter.parent().contents();
    lastParentOfFooter.replaceWith(footerParentContents);
  }
```

yamgent commented on lib/Page.js:
```diff
@@ -62,10 +63,55 @@ function calculateNewBaseUrl(filePath, root, lookUp) {
+  });
+}
+
+function removeFooterParentTags(pageData) {
```
`unwrapFooter` might sound better.

yamgent commented on docs/userGuide/includingContents.md:
````diff
@@ -62,6 +62,12 @@
         site.json
       some.md (it include a file in site A, <include src="./A/other.md">)
     ```
+    Path guide:
````
`You may reference files from external directories using the following prefixes:`

yamgent commented on lib/Site.js:
```diff
@@ -125,7 +125,7 @@ function getGlyphiconsMap() {
   return glyphiconsClasses.reduce((glyphiconsMap, glyphiconClass) => {
     const name = glyphiconClass.replace(/-/g, '_');
-    const html = `<span class='glyphicon ${glyphiconClass}' aria-hidden='true'></span>`;
+    const html = `<span class='glyphicon ${glyphiconClass}' aria-hidden='true'>&#8203</span>`;
```
I think you might want to create a new issue for this in order to explain why this change is necessary.

Chng-Zhi-Xuan replied:
This "bug" appeared when my implementation used `Cheerio` after `markbinder.renderFile`. It wouldn't have appeared otherwise.

Will clarify further in commit message.

acjh replied:
Must it be done after `markbinder.renderFile`?

Chng-Zhi-Xuan replied:
As not all the html elements / tags has been added to the pageData, I did not attempt to do it before .renderFile.  

Will experiment tomorrow.

acjh replied:
- Why does it need all the html elements / tags to have been added?
- Can we assume/limit that footer can only be in addressable pages?
- Can I clarify that footer must be explicitly included?

acjh replied:
- Shall we specify `footer` in front matter similar to what [Jekyll](https://jekyllrb.com/docs/frontmatter/) does with `layout`?
  - This allows separation from the content.
    ```html
    <frontmatter>
    title: MarkBind
    keywords: mark, markdown, markbind, fragment, reusable, CLI
    footer: footer.md
    </frontmatter>
    ```
  - Let's require file extension ([Assemble](http://assemble.io/docs/YAML-front-matter.html) does too).
  - We can assume it's relative to `_markbind/layouts`.

acjh replied:
Moved to https://github.com/MarkBind/markbind/pull/238#issuecomment-391683933 for extended discussion.

Chng-Zhi-Xuan commented:
- Changes requested has been implemented.

acjh commented:
Continued from https://github.com/MarkBind/markbind/pull/238#discussion_r190535789

> As not all the html elements / tags has been added to the pageData, I did not attempt to do it before .renderFile.
>
> Will experiment tomorrow.

- Why does it need all the html elements / tags to have been added?
- Can we assume/limit that footer can only be in addressable pages?
- Can I clarify that footer must be explicitly included?
- Shall we specify `footer` in front matter similar to what [Jekyll](https://jekyllrb.com/docs/frontmatter/) does with `layout`?
  - This allows separation from the content.
    ```html
    <frontmatter>
    title: MarkBind
    keywords: mark, markdown, markbind, fragment, reusable, CLI
    footer: footer.md
    </frontmatter>
    ```
  - Let's require file extension ([Assemble](http://assemble.io/docs/YAML-front-matter.html) does too).
  - We can assume it's relative to `_markbind/layouts`.

Chng-Zhi-Xuan commented:
> - Why does it need all the html elements / tags to have been added?
> - Can we assume/limit that footer can only be in addressable pages?
> - Can I clarify that footer must be explicitly included?
> - Shall we specify footer in front matter similar to what Jekyll does with layout?

-  I assumed wrongly that `Cheerio` would misbehave if there were elements that weren't html when parsed into its `load` method. <br> I shifted my function calls to after `removeFrontMatter` and before `markbinder.resolveBaseUrl`. It works as per normal, and without the `&#8203` spacing fix for the glyphicons. 🎉 

- Yes, as footers won't be rendered on non-addressable pages. 
- Footers are optional, user has to explicitly include a footer in each page for that page to have it.  However, we encourage it for the landing page (from comments within `footer.md`). Where a link to `MarkBind` can be given as some form of acknowledgement / advertisement (requested by @damithc) 

- This looks to be a cleaner implementation as well, without the User manually having to insert / include a footer at the end of the source file. 👍

Chng-Zhi-Xuan commented:
**Update**
- Change `footer.md` file directory to be in `_markbind/footers`
- Change including footers at end of file to be specified within `Front Matter`
- Add `insertTemplate(pageData)` Page method for to insert templates specified within `Front Matter`

**Note**
- Discussion above had a suggestion to follow `Jekyll`'s implementation of `layouts` but the scope was discovered to be too large for this PR. A similar and smaller scoped solution was implemented instead.

acjh commented:
> Would like to clarify about requiring file extension. Do you mean restrict file extensions for footer files to be either .md or .html?

Nope. Jekyll requires file name without the fileextension. Assemble requires the file extension.

acjh commented:
> - Discussion above had a suggestion to follow Jekyll's implementation of `layouts` but the scope was discovered to be too large for this PR. A similar and smaller scoped solution was implemented instead.

Yeah, I just meant to specify footer in front matter.

yamgent commented on lib/Page.js:
```diff
@@ -136,6 +187,24 @@ Page.prototype.removeFrontMatter = function (includedPage) {
+    const specifiedFooter = this.frontMatter.footer;
+    const footerFilePath = path.join(this.rootPath, FOOTERS_FOLDER_PATH, specifiedFooter);
+    if (!utils.fileExists(footerFilePath)) {
+      throw new Error(`file ${specifiedFooter} does not exist in ${FOOTERS_FOLDER_PATH}`);
```
Start the error message with capital letter: **F**ile

acjh commented on lib/markbind/lib/parser.js:
```diff
@@ -121,7 +121,9 @@ Parser.prototype._preprocess = function (node, context, config) {
-    filePath = isUrl ? element.attribs.src : path.resolve(path.dirname(context.cwf), includeSrc.path);
+    filePath = isUrl
+      ? element.attribs.src
+      : decodeURIComponent(path.resolve(path.dirname(context.cwf), includeSrc.path));
```
Rather, should we only decode `includeSrc.path` since this is meant to counter `url.parse`?

Chng-Zhi-Xuan replied:
I agree with @acjh , since `includeSrc.path` is the only element that contains the encoded characters.

yamgent replied:
Let's do that then

Chng-Zhi-Xuan commented:
**Update**
- Changes requested

Chng-Zhi-Xuan commented on docs/userGuide/contentAuthoring.md:
```diff
@@ -55,6 +55,52 @@ The HTML output will be:
+- Specified footer file in [front matter](#front-matter) overrides all inline footers in the same page.
+- Inline footers need to be the last element in a document.
+- If there is more than one inline footer in a page, only the last inline footer will be rendered.
+- Inline footers should not be nested in other components or html tags, otherwise it will not be rendered properly. 
```
- There could be multiple footers on the same nesting level. `unwrapFooter` only destroys the last footer's parents and bring it to the top. If there are multiple footers, they will all be rendered.

acjh replied:
How does this affect multiple footers? Doesn't `removeDuplicateFooters` remove them?

Chng-Zhi-Xuan replied:
- I've misread your comment, I thought you were referring to `line 101`, when you were referring to `line 102`. The comment should be rephrased as:
 
  `Footers should not be nested in other components, as the components will be removed for the footer to be rendered correctly.` 

  I wanted to convey the information that `unwrapFooter` removes any components that houses the last footer, so the footer can be "freed".

acjh replied:
Removing components doesn't seem appropriate. See https://github.com/MarkBind/markbind/pull/238#discussion_r191296354:

> - I'm not sure about `lastFooterOutermostParent.replaceWith(lastFooterParentContents);`.
>   Shouldn't we do `lastFooterOutermostParent.after(lastFooter);` instead?

Anyway, I don't think this point is necessary.

Chng-Zhi-Xuan commented on docs/userGuide/contentAuthoring.md:
```diff
@@ -55,6 +55,52 @@ The HTML output will be:
+- Inline footers need to be the last element in a document.
+- If there is more than one inline footer in a page, only the last inline footer will be rendered.
+- Inline footers should not be nested in other components or html tags, otherwise it will not be rendered properly. 
+
```
I thought about that, but the `Footers` section seem to flow better when placed under `General writing guide`

acjh replied:
`General writing guide` applies to non-localised writing with no context.

It seems apt to describe `Footers` at the end.

Let's move **userGuideSections.md** to **_markbind/footers/userGuideSections.md**:

```html
<footer>
  <markdown class="website-content">
## Sections
- <a href="{{baseUrl}}/userGuide/userQuickStart.html">Quick Start</a>
- <a href="{{baseUrl}}/userGuide/developingASite.html">Developing a Site</a>
  - <a href="{{baseUrl}}/userGuide/siteConfiguration.html">Site Configuration</a>
- <a href="{{baseUrl}}/userGuide/contentAuthoring.html">Content Authoring</a>
  - <a href="{{baseUrl}}/userGuide/includingContents.html">Including Contents</a>
- Deploying a Site
  - <a href="{{baseUrl}}/userGuide/ghpagesDeployment.html">Github Pages Deployment</a>
  - <a href="{{baseUrl}}/userGuide/netlifyPreview.html">Preview site on Netlify</a>
  </markdown>
</footer>
```

Chng-Zhi-Xuan commented on lib/Page.js:
```diff
@@ -10,8 +10,11 @@ const Promise = require('bluebird');
 const MarkBind = require('./markbind/lib/parser');
+const utils = require('./markbind/lib/utils');
 
+const FLEX_DIV_HTML = '<div id="flex-div">&#8203</div>\n';
```
- Might be a leftover from before shifting the footer post-processing to before `markbinder.renderFile`. Will test if it works without `&#8203`.

Chng-Zhi-Xuan replied:
**Update**
- Tested and works fine without `&#8203`.

Chng-Zhi-Xuan commented on lib/Page.js:
```diff
@@ -136,6 +187,24 @@ Page.prototype.removeFrontMatter = function (includedPage) {
+    modifiedPageData += `\n${nunjucks.renderString(footerData, pageUserDefinedVariables)}`;
+  }
+
+  return modifiedPageData;
```
- Method was designed initially to house potential future templates to be inserted via specifying in `Front Matter`
- Wouldn't `fs.readFileSync` throw an error if the path is invalid?

acjh replied:
> - Method was designed initially to house potential future templates to be inserted via specifying in `Front Matter`

Doesn't seem appropriate at this point.

> - Wouldn't `fs.readFileSync` throw an error if the path is invalid?

Yes, with an error no less informative than what is currently thrown.
Especially since we don't intend to recover from it.

See https://nodejs.org/api/fs.html#fs_fs_exists_path_callback:

>> Using `fs.exists()` to check for the existence of a file before calling `fs.open()`, `fs.readFile()` or `fs.writeFile()` is not recommended. Doing so introduces a race condition, since other processes may change the file's state between the two calls. Instead, user code should open/read/write the file directly and handle the error raised if the file does not exist.

acjh commented on asset/css/markbind.css:
```diff
@@ -40,3 +40,26 @@ blockquote {
+    flex-shrink: 0;
+    margin-left: -15px;
+    margin-right: -15px;
+    margin-top: 40px;
```
Hmm, that article was from 2016. Still, good to include references in PR description.

I couldn't reproduce the reason for the workarounds in Safari and Chrome, on Mac.
This seems sufficient:

```css
footer {
    background-color: #f5f5f5;
    color: dimgrey;
    margin-left: -15px;
    margin-right: -15px;
    padding: 10px 0;
}

#app {
    display: flex;
    flex-direction: column;
    min-height: 100vh;
}

#flex-div {
    flex: 1;
}
```

`#flex-div` could use a more informative name, but I don't have anything in mind.

Chng-Zhi-Xuan commented on lib/Page.js:
```diff
@@ -62,10 +65,58 @@ function calculateNewBaseUrl(filePath, root, lookUp) {
+
+  return pageData;
+}
+
```
- About `lastFooterOutermostParent.after(lastFooter);`
  Shifting footer to outside the component can be a (less destructive) alternate implementation. We would just need to remove the footer within the parent. Will update user guide to reflect this change.

Chng-Zhi-Xuan commented:
**Update**
- Ready for review

acjh commented on asset/css/markbind.css:
```diff
@@ -40,3 +40,24 @@ blockquote {
+footer {
+    background-color: #f5f5f5;
+    color: dimgrey;
+    flex-shrink: 0;
```
`flex-shrink: 0;` doesn't seem necessary, as discussed in https://github.com/MarkBind/markbind/pull/238#discussion_r191289525.

acjh commented on docs/userGuide/contentAuthoring.md:
````diff
@@ -280,6 +280,55 @@ Front matter can also be included from a separate file, just like how content ca
+After which, you must specify the file within [front matter](#front-matter)'s `footer` attribute to render the footer.
+
+```html
+<!-- In the file you want to include a footer -->
````
"In the file `that` you want to include a footer"

acjh commented on docs/userGuide/contentAuthoring.md:
````diff
@@ -280,6 +280,55 @@ Front matter can also be included from a separate file, just like how content ca
+```
+
+Note:
+- Specified footer file in [front matter](#front-matter) overrides all inline footers in the same page.
````
From https://github.com/MarkBind/markbind/pull/238#discussion_r191292696:

> "Footer file specified"

yamgent commented on docs/userGuide/contentAuthoring.md:
````diff
@@ -280,6 +280,55 @@ Front matter can also be included from a separate file, just like how content ca
+```html
+<!-- In _markbind/footers/myFooterFile.md -->
+<footer>
+    This is my own footer!
````
We use two spaces for indentation.

acjh replied:
Good catch.

yamgent commented on docs/userGuide/contentAuthoring.md:
````diff
@@ -280,6 +280,55 @@ Front matter can also be included from a separate file, just like how content ca
+```html
+<!-- In the file you want to include a footer -->
+<frontmatter>
+footer: myFooterFile.md
````
Should we add two spaces indentation to make it consistent with the Front Matter section?

acjh replied:
Yes.

yamgent commented on docs/userGuide/contentAuthoring.md:
```diff
@@ -280,6 +280,55 @@ Front matter can also be included from a separate file, just like how content ca
 
 </div>
+
+<include src="../_markbind/footers/userGuideSections.md" />
```
Think we should use `<frontmatter>` instead of using `<include>`? :P

acjh replied:
Sure.

yamgent commented on docs/userGuide/includingContents.md:
````diff
@@ -62,6 +62,13 @@
     ```
+    
+    You may reference files from external directories using the following prefixes:
+    ```html
````
Don't think the html annotation is necessary for this code block (there's no HTML code inside).

acjh replied:
Yeah, similar to the code block just before this.

acjh commented on lib/Page.js:
```diff
@@ -62,6 +65,28 @@ function calculateNewBaseUrl(filePath, root, lookUp) {
+    lastFooterOutermostParent.after(lastFooter);
+  }
+  // Insert flex div before last footer
+  if (lastFooter.before().attr('id') !== 'flex-div') {
```
If we're defining `FLEX_DIV_HTML`, then we should define `FLEX_DIV_ID` too.

acjh commented on lib/Site.js:
```diff
@@ -60,8 +61,20 @@ const SITE_CONFIG_DEFAULT = {
+
 const INDEX_MARKDOWN_DEFAULT = '<frontmatter>\n'
   + 'title: "Hello World"\n'
+  + 'footer: footer.md\n'
```
Let's indent `title` and `footer`.

acjh commented on docs/userGuide/contentAuthoring.md:
```diff
@@ -280,6 +280,55 @@ Front matter can also be included from a separate file, just like how content ca
+- Specified footer file in [front matter](#front-matter) overrides all inline footers in the same page.
+- Inline footers need to be the last element in a page.
+- If there is more than one inline footer in a page, only the last inline footer will be rendered.
+- Inline footers should not be nested in other components or html tags. If the footer is found inside, it will be shifted outside to be rendered properly.
```
"- `Footers` should not be nested in other components or `HTML` tags. If ~`the footer is`~ found inside, `they` will be shifted outside to be rendered properly."

Chng-Zhi-Xuan commented:
**Update**
- Ready for review

acjh commented on lib/Site.js:
```diff
@@ -60,8 +61,20 @@ const SITE_CONFIG_DEFAULT = {
 };
 
+const FOOTER_DEFAULT = '<footer>\n'
+  + '  <div style="text-align:center;">\n'
```
Let's use Bootstrap's `class="text-center"`.

acjh commented on lib/Page.js:
```diff
@@ -62,6 +66,28 @@ function calculateNewBaseUrl(filePath, root, lookUp) {
+    lastFooterOutermostParent.after(lastFooter);
+  }
+  // Insert flex div before last footer
+  if (lastFooter.before().attr('id') !== FLEX_DIV_ID) {
```
Should be `.prev()` instead of `.before()`. Can test by manually inserting `FLEX_DIV_HTML`.

acjh commented on lib/Page.js:
```diff
@@ -11,6 +11,10 @@ const FsUtil = require('./util/fsUtil');
 
+const FOOTERS_FOLDER_PATH = '_markbind/footers/';
+
+const FLEX_DIV_HTML = '<div id="flex-div"></div>\n';
```
Trailing  `\n` doesn't seem necessary.

Chng-Zhi-Xuan replied:
Assumed that it would make the html output more readable. But everything is handled nicely with `htmlBeautify` 👍. Tested without `\n` and looks fine.

acjh commented on lib/Page.js:
```diff
@@ -11,6 +11,10 @@ const FsUtil = require('./util/fsUtil');
 const logger = require('./util/logger');
 const MarkBind = require('./markbind/lib/parser');
 
+const FOOTERS_FOLDER_PATH = '_markbind/footers/';
```
Trailing `/` doesn't seem necessary.

acjh commented:
> One minor nit

Where at? 🙊

yamgent commented:
> Where at?

Updated my review, forgotten to add that comment.

BTW, to make sure that the footer generates correctly, I think we can also try out the footer frontmatter in our 'test_site' test.

Chng-Zhi-Xuan commented:
**Update**
- Ready for review

yamgent commented:
@Chng-Zhi-Xuan Now might be a good time to squash the commits.

acjh commented on lib/Site.js:
```diff
@@ -60,8 +61,20 @@ const SITE_CONFIG_DEFAULT = {
 
+const FOOTER_DEFAULT = '<footer>\n'
+  + '  <div class="text-center">\n'
+  + '    This is a dynamic height footer that support variables {{glyphicon_tags}} '
```
"supports"

acjh commented on lib/Page.js:
```diff
@@ -136,6 +162,18 @@ Page.prototype.removeFrontMatter = function (includedPage) {
   return $.html();
 };
 
+Page.prototype.insertFooter = function (pageData) {
```
Add jsdoc.

acjh commented on test/test_site/_markbind/footers/footer.md:
```diff
@@ -0,0 +1,5 @@
+<footer>
+  <div class="text-center">
+    This is a dynamic height footer that support variables {{glyphicon_tags}} and markdown <md>:smile:</md>!
```
"supports"
